### ISSUE TITLE
Nested lists might cause stackoverflow

### ISSUE THREAD

**Issue by jk-idealo** [Nylle/JavaFixture#89]
I observed weird behaviour when using a nested list in a fixture, and I found it very hard to reproduce.

We use JF in a `@DataMongoTest` with `@Testcontainers` which looks like this:
```java
@DataMongoTest
@Testcontainers
class MyRepositoryIT {

    @Container
    static MongoDBContainer mongoDbContainer = new MongoDBContainer(DockerImageName.parse("mongo:4.4.13"));

    @DynamicPropertySource
    static void setMongoContainer(DynamicPropertyRegistry registry) {
        mongoDbContainer.start();
        registry.add("spring.data.mongodb.uri", () -> mongoDbContainer.getReplicaSetUrl());
    }

    @Autowired
    MyRepository myRepository;

    @TestWithFixture
    void save(List<SomeObject> anotherFixture) {
        var myEntity = fixture().create(MyEntityWithNestedList.class);

        myRepository.save(myEntity);
    }
}
```

The `MyEntity`-class looks like this:
```java
@Getter
@Setter
@NoArgsConstructor
@Document
public class MyEntity {

    @Id
    @Setter(AccessLevel.PRIVATE)
    private Integer id;

    private List<List<SomeObject>> listOfLists;

    public MyEntity(List<List<SomeObject>> listOfLists) {
        this.listOfLists = listOfLists;
    }
}
```

When running the test, I get a StackOverflowError like this:
```
java.lang.StackOverflowError
	at java.base/java.util.WeakHashMap.get(WeakHashMap.java:402)
	at org.springframework.data.mapping.model.SimpleTypeHolder.isSimpleType(SimpleTypeHolder.java:147)
	at org.springframework.data.convert.CustomConversions.isSimpleType(CustomConversions.java:163)
	at org.springframework.data.mongodb.core.convert.MappingMongoConverter.writeCollectionInternal(MappingMongoConverter.java:1109)
	at org.springframework.data.mongodb.core.convert.MappingMongoConverter.writeCollectionInternal(MappingMongoConverter.java:1115)
	at org.springframework.data.mongodb.core.convert.MappingMongoConverter.writeCollectionInternal(MappingMongoConverter.java:1115)
...
```

When **debugging**, it turns out, that the listOfLists is **not** `List<List<SomeObject>>` but a List<List<List<List...>>> with infinite depth.

I tried to reproduce this without DataMongo and Testcontainers in the JavaFixture-project:
```java
    @TestWithFixture
    void xxx(List<TestObject> anotherFixture) {
        var testObject = fixture().create(TestObjectWithNestedMapsAndLists.class);

        var integer = testObject.getNestedList().get(0).get(0).getIntegers().get(0);
        assertThat(integer).isNotZero();
    }
```
And here comes the first "weird" observation: if I **run** this test, it's green, but if I **debug** this test, I get `java.lang.ClassCastException: class java.util.ArrayList cannot be cast to class com.github.nylle.javafixture.testobjects.TestObject (java.util.ArrayList is in module java.base of loader 'bootstrap'; com.github.nylle.javafixture.testobjects.TestObject is in unnamed module of loader 'app')` and I can see in my evaluation-window, that I indeed get an infinetly nested List<List<List<List...>>> just like in my productive service.

**Question 1: Why does this only happen in debug, but not during a normal run?**

Then I played around a bit and removed the `@TestWithFixture`-annotation:
```java
    @Test
    void xxx() {
        var testObject = fixture().create(TestObjectWithNestedMapsAndLists.class);

        var integer = testObject.getNestedList().get(0).get(0).getIntegers().get(0);
        assertThat(integer).isNotZero();
    }
```
Now, both **run** and **debug** work just fine and are green. It seems, the combination of `@TestWithFixture`and an inner `fixture()` are somehow influencing each other. 

**Question 2: How does `@TestWithFixture` influence the inner `fixture()`?**

With that in mind, I went back to my productive service's test and removed the annotation there as well:
```java
    @TestWithFixture
    void save() {
        var myEntity = fixture().create(MyEntityWithNestedList.class);

        myRepository.save(myEntity);
    }
```
Unfortunately, I still get the `StackOverflowError` and when debugging I can see the infinitely nested `List`s. THis is the third weirdness.

**Question 3: Why does the "fix" (see question 2) does not help here?**

This last question is moot as long as I cannot reproduce the behaviour in proper tests. I would hope that if we answer the first two questions, this last problem might be resolved. If not, I would dig back into it in a second iteration.

**Comment by akutschera** [comment]
That's an interesting one.
I can verify that this behaviour is also observable on my machine, even though I cannot write a deterministic test (yet).

> **Question 1: Why does this only happen in debug, but not during a normal run?**

It doesn't. I've had the exact opposite behaviour. Sometimes it fails in a non-debug run and succeeds in debug. So that's not the issue.

> **Question 2: How does `@TestWithFixture` influence the inner `fixture()`?**

It doesn't. I can reproduce it with a normal JUnit5 test, so the fixture-injection in the test is not necessary.

I added a ton of debug messages and found out that we have a caching problem. For some reason we generate the same hash value for a `List<List<TestObject>>` and a `TestObject`. That leads to a reflection-injection of a List where we want to inject a TestObject. Since we inject a random number of elements into a list, this sometimes leads to a StackOverflowError and sometimes to a ClassCastException. Simply put, we generate the wrong object tree.

**Comment by Nylle** [comment]
I had a suspicion it's the caching. That's why I thought the `@TestWithFixture`-annotation is involved. Thanks for clearing that up! 

I guess, it shouldn't be too hard to improve the caching, but what really bothers me is that we can't get a reliably red test.

**Comment by akutschera** [comment]
From the [Javadoc of hashCode](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Object.html#hashCode()) the contract for hashCode says that

-  [the result] need not remain consistent from one execution of an application to another execution of the same application.
- It is *not* required that if two objects are unequal according to the equals method, then calling the hashCode method on each of the two objects must produce distinct integer results.

So this means that it could even be a different hashCode on another JVM. So I guess we should not rely on the hashCode as a unique key for our cache (unless we implement it ourselves which I don't think we should do).

Instead, if we use the `SpecimenType<?>` as the key, we would get a cache where we can have different values for a `List<List<TestObject>>` and for a `TestObject` (as long as those are not `equal`).

So I refactored the cache and the `SpecimentType.equals` method, all tests are green and two different SpecimentTypes will not have the same cache value (unless there's a bug in the `SpecimenType.equals` method)

**Comment by Nylle** [comment]
I have a question. As I have no access to the branch, can you try something for me?

On the current master, the following test is green:
```java
    @Test
    void xxx() {
        var sut = new Fixture(configuration);

        var actual = sut.create(new SpecimenType<List<String>>(){});

        assertThat(actual.size()).isGreaterThanOrEqualTo(2);
        assertThat(actual.get(0)).isNotEqualTo(actual.get(1));
    }
```

If I remember correctly, the need for introducing caching was not for performance reasons, but as the simplest available way to avoid infinite nesting, if a class has a property of its own type (or more complex, a class that has children which reference back to their parent). Caching would ensure that no infinite instances of those classes would be created but the cached one being re-used.
However, this (somehow) didn't happen for primitives. Those were still random, that means a list of string would generate multiple different strings, as proven in the test above.

Would this behaviour still be the case with the new caching? I fear, that now Strings would be cached, too and that current behaviour would change.

**Comment by akutschera** [comment]
Yes, this test is still green because we don't cache PrimitiveSpecimen. I did not change **when** an object is cached, I only changed the key in the cache map.
I added the second assertion to an already existing test.

**Comment by Nylle** [comment]
Fixed in version [2.9.6](https://github.com/Nylle/JavaFixture/releases/tag/2.9.6).
